### PR TITLE
Improve corrupted data check

### PR DESCRIPTION
**Pull request by sabramkin**

### CODE REVIEW

File: tests/corrupted_stamp_test.cpp [hunk 1]
```diff
@@ -125,6 +124,7 @@ namespace tests {
 void test_random_valid_headers(unsigned int number) {
 	std::default_random_engine random_generator;
 
+	std::discrete_distribution<uint8_t> version_distribution({10, 90}); // version=0: 10%, version=1: 90%
```
Comment (shaitan) [inline tests/corrupted_stamp_test.cpp]: Why is it such relation (10% vs 90%)?

Comment (sabramkin) [reply tests/corrupted_stamp_test.cpp]: I think that uniform distribution for version doesn't behave as true. I don't insist.

File: example/eblob_backend.cpp [hunk 2]
```diff
@@ -66,41 +66,54 @@ int blob_check_corrupted_stamp(void *buffer, size_t buffer_size) {
...
+		// dnet_ext_list_hdr signature ok
 		return -EILSEQ;
 	}
 
 	return 0;
 }
 
 int blob_read_and_check_stamp(const eblob_backend_config *c,
-                              const dnet_time *timestamp,
+                              const struct dnet_time *timestamp,
```
Comment (shaitan) [inline example/eblob_backend.cpp]: `struct` isn't necessary

Comment (sabramkin) [reply example/eblob_backend.cpp]: Oops, I manually undone some changes and copy-pasted it. Thanks.

File: example/eblob_backend.cpp [hunk 3]
```diff
@@ -85,22 +101,19 @@ int blob_read_and_check_stamp(const eblob_backend_config *c,
                               int fd,
                               uint64_t data_offset,
                               uint64_t data_size) {
-	std::array<char, sizeof(eblob_disk_control) + sizeof(dnet_ext_list_hdr)> stamp;
-
-	if (data_size < stamp.size()) {
+	if (timestamp && timestamp->tsec > DNET_SERVER_SEND_BUGFIX_TIMESTAMP) {
```
Comment (shaitan) [inline example/eblob_backend.cpp]: Why was it changed from `!timestamp || timestamp->tsec > DNET_SERVER_SEND_BUGFIX_TIMESTAMP`?

File: tests/corrupted_stamp_test.cpp [hunk 4]
```diff
@@ -76,82 +77,89 @@ std::vector<headers_test_case> marginal_cases = {
 	// data_size <= disk_size
 	{0, 1, 100, 99, DNET_EXT_VERSION_V1, 0, {0}, {0}},
 	{0, 1 << 9, 100, 42, DNET_EXT_VERSION_V1, 0, {0}, {0}},
-
```
Comment (shaitan) [inline tests/corrupted_stamp_test.cpp]: Why was it removed?

File: tests/corrupted_stamp_test.cpp [hunk 5]
```diff
@@ -76,82 +77,89 @@ std::vector<headers_test_case> marginal_cases = {
...
-	// version != 1
-	{0, 1, 1, 99, 0, 0, {0}, {0}},
-	{0, 1, 1, 99, 2, 0, {0}, {0}},
+
+	// version isn't equal to DNET_EXT_VERSION_V1 or DNET_EXT_VERSION_FIRST
+	{0, BLOB_DISK_CTL_EXTHDR, 1, 99, 2, 0, {0}, {0}},
 	// timestamp out of range.
-	{0, 1, 100, 101, DNET_EXT_VERSION_V1, DNET_SERVER_SEND_BUGFIX_TIMESTAMP + 1, {0}, {0}},
-	{0, 1, 100, 101, DNET_EXT_VERSION_V1, DNET_SERVER_SEND_BUGFIX_TIMESTAMP + 42, {0}, {0}},
-	{0, 1, 100, 101, DNET_EXT_VERSION_V1, DNET_SERVER_SEND_BUGFIX_TIMESTAMP + 100500, {0}, {0}},
```
Comment (shaitan) [inline tests/corrupted_stamp_test.cpp]: Why were they removed?

File: tests/corrupted_stamp_test.cpp [hunk 6]
```diff
@@ -76,82 +77,89 @@ std::vector<headers_test_case> marginal_cases = {
...
-	{0, 1, 42, 42, DNET_EXT_VERSION_V1, 0, {0, 0, 1}, {0, 0}},
-	{0, 1, 42, 42, DNET_EXT_VERSION_V1, 0, {0, 0, 0}, {1, 0}},
-	{0, 1, 42, 42, DNET_EXT_VERSION_V1, 0, {0, 0, 0}, {0, 1}},
-	{0, 1, 42, 42, DNET_EXT_VERSION_V1, 0, {0, 0, 0}, {1, 1}},
-	{0, 1, 42, 42, DNET_EXT_VERSION_V1, 0, {1, 1, 1}, {0, 0}},
-	{0, 1, 42, 42, DNET_EXT_VERSION_V1, 0, {1, 0, 1}, {0, 0}},
-	{0, 1, 42, 42, DNET_EXT_VERSION_V1, 0, {1, 1, 1}, {1, 0}},
-	{0, 1, 42, 42, DNET_EXT_VERSION_V1, 0, {1, 1, 1}, {0, 1}},
-	{0, 1, 42, 42, DNET_EXT_VERSION_V1, 0, {1, 0, 1}, {1, 0}},
-	{0, 1, 42, 42, DNET_EXT_VERSION_V1, 0, {1, 0, 1}, {0, 1}}
```
Comment (shaitan) [inline tests/corrupted_stamp_test.cpp]: Why were they removed?

File: tests/corrupted_stamp_test.cpp [hunk 7]
```diff
@@ -239,13 +252,14 @@ void test_read_corrupted_stamp(ioremap::elliptics::newapi::session &s, const dne
...
-		                         constants::HEADERS_SIZE + to_add - to_remove,
+		                         get_headers_size(test_case.dc_flags),
+		                         get_headers_size(test_case.dc_flags) + to_add - to_remove,
 		                         std::move(async)};
 	};
 
 	std::vector<test_write_result> write_results;
 
-	size_t id = 0;
+	int id = 0;
```
Comment (shaitan) [inline tests/corrupted_stamp_test.cpp]: Why is it `int` now?

File: tests/corrupted_stamp_test.cpp [hunk 8]
```diff
@@ -350,25 +366,25 @@ void test_read_corrupted_stamp(ioremap::elliptics::newapi::session &s, const dne
 		// Different offests and data size requests.
 		//
 		{
-			auto async = s.read_data(write_result.key, 11, 101);
+			auto async = s.read_data(write_result.key, 11, 81);
```
Comment (shaitan) [inline tests/corrupted_stamp_test.cpp]: Why was it changed?

File: tests/corrupted_stamp_test.cpp (hunk 1 above)
Comment (shaitan) [reply tests/corrupted_stamp_test.cpp]: Why should the test have this distribution?

Comment (sabramkin) [reply tests/corrupted_stamp_test.cpp]: Test cases are random generated. There's no guarantee that they each time will be generated to cover all the functional cases. False-positive test pass can occur. The main cases must be tested more properly, because false-positive result is more critical for them.

File: example/eblob_backend.cpp (hunk 3 above)
Comment (sabramkin) [reply example/eblob_backend.cpp]: No timestamp = we cannot decide to prevent check - isn't it?

But I see that `blob_read_and_check_stamp` never called with `timestamp == nullptr`, so we can remove any check for empty ts.

Comment (sabramkin) [reply example/eblob_backend.cpp]: I reverted that change.

File: example/eblob_backend.cpp [hunk 9]
```diff
@@ -66,41 +66,54 @@ int blob_check_corrupted_stamp(void *buffer, size_t buffer_size) {
...
 
 int blob_read_and_check_stamp(const eblob_backend_config *c,
-                              const dnet_time *timestamp,
+                              const struct dnet_time *timestamp,
                               int fd,
                               uint64_t data_offset,
                               uint64_t data_size) {
-	std::array<char, sizeof(eblob_disk_control) + sizeof(dnet_ext_list_hdr)> stamp;
-
-	if (data_size < stamp.size()) {
```
[the remark this answers is not in the document]
Comment (sabramkin) [reply example/eblob_backend.cpp]: To prevent reading? Ok.

File: tests/corrupted_stamp_test.cpp (hunk 4 above)
Comment (sabramkin) [reply tests/corrupted_stamp_test.cpp]: To group cases about disk_size. Too much separators. I can revert.

File: tests/corrupted_stamp_test.cpp (hunk 5 above)
Comment (sabramkin) [reply tests/corrupted_stamp_test.cpp]: No difference between `DNET_SERVER_SEND_BUGFIX_TIMESTAMP + 1` and `DNET_SERVER_SEND_BUGFIX_TIMESTAMP + 42`. It's the only case enough: `time` is greater than `DNET_SERVER_SEND_BUGFIX_TIMESTAMP`. No benefit from repeating of the equal cases. Just the run-time grows. Tests must run fast.

File: tests/corrupted_stamp_test.cpp (hunk 6 above)
Comment (sabramkin) [reply tests/corrupted_stamp_test.cpp]: It's very simple case: "all pads must be zero". A couple of tests is enough. What if we'll have 100500 pad-fields? No benefit from repeating of the equal cases. Just the run-time grows. Tests must run fast.

File: example/eblob_backend.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (shaitan) [reply example/eblob_backend.cpp]: Yep

File: tests/corrupted_stamp_test.cpp (hunk 7 above)
Comment (sabramkin) [reply tests/corrupted_stamp_test.cpp]: Thanks, I'll revert.
I reverted some changes and occasionally pasted it.

File: tests/corrupted_stamp_test.cpp (hunk 8 above)
Comment (sabramkin) [reply tests/corrupted_stamp_test.cpp]: `sizeof(dc)` = 96, `sizeof(ehdr)` = ~40.

96 + 40 < 11 + 101, so the test was working.

But now we have the case of `dc` only. Tests with old values became broken.

File: tests/corrupted_stamp_test.cpp (hunk 6 above)
Comment (sabramkin) [reply tests/corrupted_stamp_test.cpp]: But I can revert.

File: tests/corrupted_stamp_test.cpp (hunk 5 above)
Comment (sabramkin) [reply tests/corrupted_stamp_test.cpp]: But I can revert.

File: tests/corrupted_stamp_test.cpp (hunk 4 above)
Comment (shaitan) [reply tests/corrupted_stamp_test.cpp]: ok

File: tests/corrupted_stamp_test.cpp (hunk 8 above)
Comment (shaitan) [reply tests/corrupted_stamp_test.cpp]: `sizeof(ehdr) = 48`. Please add a comment why `81` is used here.

File: example/eblob_backend.cpp [hunk 10]
```diff
@@ -66,14 +66,30 @@ int blob_check_corrupted_stamp(void *buffer, size_t buffer_size) {
 		       dc.position == 0;
 	};
 
+	if (!check_dc_header(*dc)) {
+		// eblob_disk_control signature check failed
```
Comment (shaitan) [inline example/eblob_backend.cpp]: I would rephrase this and comments below to: `eblob_disk_control signature wasn't found the key looks normal`
